File: web/frontend/static/js/sessions.js
/**
 * The recordings page — every session that has been captured.
 *
 * Its own entry point rather than part of `main.js`: this page has no WebSocket, no transcript
 * store, and no assistant, and loading that machinery to render a list would mean a page that
 * cannot open when the pipeline is unhealthy — exactly when someone wants to retrieve a transcript.
 */

import { ExportDialog } from "./components/export-dialog.js";
import { $, el, setText, toggle } from "./core/dom.js";
import { duration as formatDuration } from "./core/format.js";
import { exportJob } from "./stores/export.js";
import { api } from "./transport/api.js";


/**
 * Named the way the transcript pane names them: "Final" means something to a reader and
 * "revision 1" does not. Anything beyond the two the application produces today falls back to a
 * number rather than being hidden.
 */
const REVISION_LABELS = { 0: "Live", 1: "Final" };
const revisionLabel = (revision) => REVISION_LABELS[revision] ?? `Version ${revision}`;

const FORMATS = [
  ["markdown", "Markdown"],
  ["text", "Plain text"],
  ["srt", "SRT"],
  ["vtt", "WebVTT"],
  ["json", "JSON"],
];

function when(iso) {
  if (!iso) return "";
  const date = new Date(iso);
  return Number.isNaN(date.getTime())
    ? ""
    : date.toLocaleString(undefined, {
        day: "numeric",
        month: "short",
        year: "numeric",
        hour: "2-digit",
        minute: "2-digit",
      });
}

/** The three things a recording can hold, in a fixed order so a column of rows scans. */
const MEDIA = [
  ["video", "Video"],
  ["audio", "Audio"],
  ["transcript", "Transcript"],
];

/**
 * What a chip's tooltip says.
 *
 * Audio gets a sentence of its own because "yes" has two meanings worth telling apart: the
 * separate recording is still on disk, or a successful transcription pass deleted it and the sound
 * now lives in the video. Both are audio; only one of them can be transcribed again.
 */
function chipTitle(key, label, present, media) {
  if (!present) return `No ${label.toLowerCase()}`;
  if (key !== "audio") return `${label} is available`;
  return media.audio_file
    ? "The recorded audio is still on disk and can be transcribed again"
    : "The audio is in the video — the separate recording was removed once it was transcribed";
}

function size(bytes) {
  return bytes >= 1024 * 1024
    ? `${(bytes / 1024 / 1024).toFixed(1)} MB`
    : `${Math.max(1, Math.round(bytes / 1024))} KB`;
}

class SessionsPage {
  constructor() {
    this.list = $("[data-sessions-list]");
    this.empty = $("[data-sessions-empty]");
    this.error = $("[data-sessions-error]");
    this.directory = $("[data-sessions-directory]");
    this.runningKey = "";
    // The same window that opens when a recording stops, reached from the other end. This page has
    // no socket, so it polls while an export runs — see `openExport`.
    this.exportDialog = new ExportDialog($("[data-export]"));
  }

  /**
   * Open the export window for one session, and keep it fed.
   *
   * **This page deliberately has no WebSocket** — its own header says so: loading the transport,
   * the stores and the transcript machinery to render a list would mean a page that cannot open
   * when the pipeline is unhealthy, which is exactly when someone wants to retrieve a recording.
   * So an export's progress is polled here rather than pushed, which for a job that reports about
   * once a second is a difference nobody can see.
   */
  async openExport(session) {
    const poll = setInterval(async () => {
      if (!this.exportDialog.isOpen) return;
      try {
        const body = await api.exportStatus(session.key);
        if (body.job) exportJob.set(body.job);
      } catch {
        // A failed poll is a frame missed, not a state. The next one answers.
      }
    }, 1000);

    try {
      await this.exportDialog.show(session.key, { title: session.title });
    } finally {
      clearInterval(poll);
      await this.load();
    }
  }

  async load() {
    try {
      const body = await api.sessions();
      this.runningKey = body.running_key ?? "";
      setText(this.directory, `Stored in ${body.directory}`);
      this.render(body.sessions ?? []);
      toggle(this.error, false);
    } catch (error) {
      toggle(this.error, true);
      setText(this.error, error.message);
    }
  }

  render(sessions) {
    toggle(this.empty, sessions.length === 0);
    this.list.replaceChildren(...sessions.map((session) => this.row(session)));
  }

  row(session) {
    const running = session.key === this.runningKey;
    const media = session.media ?? {};
    const stored = media.recording_bytes
      ? size(session.size_bytes + media.recording_bytes)
      : size(session.size_bytes);
    const meta = [
      when(session.started_at),
      session.duration_seconds ? formatDuration(session.duration_seconds) : null,
      `${session.segments} segments`,
      `${session.words} words`,
      session.glossary_terms ? `${session.glossary_terms} terms` : null,
      stored,
    ].filter(Boolean);

    const children = [
      el("div", {
        className: "session__head",
        children: [
          el("h2", { className: "session__title", text: session.title }),
          // Marked rather than hidden: it is the session the user is most likely looking for, and
          // its numbers are about to change.
          running ? el("span", { className: "badge", text: "recording now" }) : null,
        ],
      }),
      el("p", { className: "session__meta", text: meta.join(" · ") }),
      this.media(media),
    ];

    if (session.problem) {
      children.push(el("p", { className: "session__problem", text: session.problem }));
    } else {
      children.push(this.actions(session, running));
    }

    return el("article", {
      className: "session",
      attrs: { role: "listitem", "data-session-key": session.key },
      children,
    });
  }

  /**
   * What this recording holds — video, audio, transcript — as three chips.
   *
   * All three are always drawn, present or not. A row that showed only what it had would need
   * reading rather than scanning, because "video · transcript" and "video · audio" have the same
   * shape and different meanings.
   */
  media(media) {
    return el("div", {
      className: "session__media",
      attrs: { "aria-label": "What this recording holds" },
      children: MEDIA.map(([key, label]) => {
        const present = Boolean(media[key]);
        return el("span", {
          className: "media-chip",
          text: label,
          attrs: { "data-present": String(present), title: chipTitle(key, label, present, media) },
        });
      }),
    });
  }

  actions(session, running) {
    const format = el("select", {
      className: "field__control field__control--compact",
      attrs: { "aria-label": `Export format for ${session.title}` },
      children: FORMATS.map(([value, label]) =>
        el("option", { text: label, attrs: { value } })
      ),
    });

    // **Only where there is a choice (D-066).** A session that transcribed live and again
    // afterwards holds two passes, and the export used to ship the newest with no way to ask for
    // the other. The control is drawn only for those sessions — a select with one option is
    // chrome that explains nothing, the same rule the transcript pane's Live/Final switch follows.
    const revisions = Array.isArray(session.revisions) ? session.revisions : [];
    const latest = revisions.length ? revisions[revisions.length - 1] : undefined;
    const version =
      revisions.length > 1
        ? el("select", {
            className: "field__control field__control--compact",
            attrs: { "aria-label": `Transcript version to export for ${session.title}` },
            children: revisions.map((revision) =>
              el("option", {
                text: revisionLabel(revision),
                attrs: { value: String(revision), ...(revision === latest ? { selected: "" } : {}) },
              })
            ),
          })
        : null;
    const chosenRevision = () => (version ? Number(version.value) : undefined);

    const download = el("a", {
      className: "button",
      text: "Export",
      attrs: {
        href: api.sessionExportUrl(session.key, "markdown", chosenRevision()),
        download: "",
      },
    });
    // The link's href follows both dropdowns, so the browser handles the download itself and the
    // filename comes from the server's own header rather than being invented here.
    const follow = () => {
      download.href = api.sessionExportUrl(session.key, format.value, chosenRevision());
    };
    format.addEventListener("change", follow);
    version?.addEventListener("change", follow);

    // **Its own button, and only where there is one.** The transcript export deliberately no
    // longer carries the conversation: a transcript is usually being handed to somebody else, and
    // one person's questions are not part of the record of the talk. They are still the user's,
    // and this is how they get them.
    const chat = session.chat_messages
      ? el("a", {
          className: "button button--quiet",
          text: "Conversation",
          attrs: {
            href: api.sessionChatUrl(session.key, "markdown"),
            download: "",
            title:
              `The ${session.chat_messages} messages you exchanged with the assistant during ` +
              "this recording, on their own. Not included in the transcript or the web app.",
          },
        })
      : null;

    const remove = el("button", {
      className: "button button--quiet",
      text: "Delete",
      attrs: { type: "button", disabled: running },
    });
    remove.addEventListener("click", () => this.remove(session));

    // Offered only when the session holds video, audio, and a transcript. The export *is* those
    // three — a player, a transcript that follows it, and questions asked against both — so a
    // button that produced two of them under the same name would disappoint quietly.
    //
    // **A button rather than a link, since D-037.** It used to be an `<a download>` that fetched
    // hundreds of megabytes with no options and no idea what it would produce. It opens the export
    // window now: the same window that opens when a recording stops, reached from the other end.
    let webapp = null;
    if (session.media?.exportable) {
      webapp = el("button", {
        className: "button button--primary",
        text: "Export…",
        attrs: {
          type: "button",
          title:
            "Preview the recording, choose its size and quality, and watch the export run. " +
            "Your own conversation is not included unless you ask for it.",
        },
      });
      webapp.addEventListener("click", () => void this.openExport(session));
    }

    return el("div", {
      className: "session__actions",
      children: [version, format, download, chat, webapp, remove].filter(Boolean),
    });
  }

  async remove(session) {
    const confirmed = window.confirm(
      `Delete "${session.title}"? The transcript, summaries, and glossary all go with it, and this cannot be undone.`
    );
    if (!confirmed) return;

    try {
      await api.deleteSession(session.key);
      await this.load();
    } catch (error) {
      toggle(this.error, true);
      setText(this.error, error.message);
    }
  }
}

new SessionsPage().load();

File: web/backend/app/routes/sessions.py
"""Past sessions — listing, reading, exporting, and deleting.

The live application only ever knows about the session currently open. Once one stops, its
transcript lives in a SQLite file and every route under ``/api/transcript`` stops answering for it,
because those read the *running* session's store. That is correct for the live view and useless for
a talk that finished ten minutes ago, which is what these routes are for.

A session is addressed by its **key** — the file's stem — never by a path. A path from a request is
a path traversal waiting to be written, and there is no reason for the caller to supply one. That
same key names the recording's folder, which is what lets the web-app export here reach the video.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException, Query, Request
from fastapi.responses import FileResponse, Response
from starlette.concurrency import run_in_threadpool

from ..services.export import DEFAULT_PRESET, PRESETS, ExportError, SourceProfile, build_webapp
from ..services.export import by_id as by_preset_id
from ..services.export import estimate as estimate_export
from ..services.export import probe as probe_media
from ..services.recording import resolve_recording
from ..services.transcript import archive
from ..services.transcript import export as render_export
from ..services.transcript import export_chat as render_chat_export
from ..services.transcript.store import TranscriptStore

router = APIRouter(prefix="/api/sessions", tags=["sessions"])


def _config(request: Request):  # noqa: ANN202 - returns AppConfig
    return request.app.state.config.resolve()


def _open(request: Request, key: str) -> TranscriptStore:
    """Open a past session read-only, or 404 with a message that says what happened."""
    path = archive.find(_config(request), key)
    if path is None:
        raise HTTPException(
            status_code=404,
            detail={
                "error": {
                    "code": "no-such-session",
                    "message": (
                        "That session is not in the session folder any more. "
                        "It may have been moved, deleted, or the folder changed in settings."
                    ),
                    "severity": "warning",
                }
            },
        )
    return TranscriptStore(path)


@router.get("")
async def list_sessions(request: Request) -> dict[str, Any]:
    """Every session on disk, newest first.

    Includes files that cannot be read, each with its reason. A session the user can see and cannot
    open is more useful than one that has silently vanished from the list.
    """
    config = _config(request)
    sessions = archive.list_sessions(config)
    running = getattr(request.app.state, "session_manager", None)
    return {
        "sessions": [session.as_dict() for session in sessions],
        "directory": str(archive.session_dir(config)),
        # So the page can mark the row that is still being written rather than presenting it as
        # finished with a duration that is about to change.
        "running_key": _running_key(running),
    }


def _running_key(manager: Any) -> str:
    """The key of the session being recorded *right now*, or empty when nothing is.

    **Gated on `is_running`, not on the store existing.** `SessionManager.store` deliberately keeps
    returning the last finished session's store so the assistant can still be asked about it
    (D-031) — so reading it alone marked every completed recording as still recording, gave it the
    "recording now" badge, and disabled its Delete button. The store answers "what can be read";
    only `is_running` answers "what is being written".
    """
    if manager is None or not getattr(manager, "is_running", False):
        return ""
    store = getattr(manager, "store", None)
    path = getattr(store, "path", None)
    return path.stem if path is not None else ""


@router.get("/{key}")
async def read_session(
    request: Request, key: str, limit: int = Query(2000, ge=1, le=20_000)
) -> dict[str, Any]:
    """One past session's transcript, summaries, and glossary."""
    store = _open(request, key)
    try:
        metadata = store.metadata()
        return {
            "key": key,
            "session": metadata.as_dict() if metadata else None,
            "stats": store.stats().as_dict(),
            # One transcription pass, not every row: a session holding both a live and a
            # post-capture pass would otherwise render the same talk twice on the page (D-022).
            "segments": [segment.as_event() for segment in store.latest_segments()[:limit]],
            "summaries": [summary.as_event() for summary in store.summaries()],
            "glossary": [term.as_event() for term in store.glossary()],
            "chat": [message.as_dict() for message in store.chat_history()],
        }
    finally:
        store.close()


@router.get("/{key}/export")
async def export_session(
    request: Request,
    key: str,
    fmt: str = Query("markdown"),
    include_chat: bool = Query(False),
) -> Response:
    """Download a past session in any of the five formats.

    **The conversation is not in it unless it is asked for.** Markdown and JSON used to carry the
    questions the user put to the assistant during the recording, because they were the two formats
    that could. But a transcript export is mostly a thing being handed to somebody else, and one
    person's half of a conversation is not part of the record of the talk. It exports on its own
    from ``/{key}/chat``; `include_chat=true` puts it back in here for anyone who wants one file.
    """
    store = _open(request, key)
    try:
        metadata = store.metadata()
        body, mime, extension = render_export(
            fmt,
            segments=store.latest_segments(),
            metadata=metadata,
            summaries=store.summaries(),
            glossary=store.glossary(),
            chat=store.chat_history() if include_chat else None,
        )
    except ValueError as exc:
        raise HTTPException(
            status_code=422, detail={"error": {"code": "unknown-format", "message": str(exc)}}
        ) from exc
    finally:
        store.close()

    stem = metadata.started_at.strftime("%Y%m%d-%H%M") if metadata else key
    return Response(
        content=body,
        media_type=mime,
        headers={"Content-Disposition": f'attachment; filename="{stem}-transcript.{extension}"'},
    )


@router.get("/{key}/chat")
async def export_chat_only(request: Request, key: str, fmt: str = Query("markdown")) -> Response:
    """Download the conversation the user had with the assistant during this recording.

    Its own document, because it belongs to the person who asked rather than to the talk. Empty is
    a valid answer and renders as a sentence saying so, rather than a 404 — "you asked nothing
    during this recording" is information, and an error code is not.
    """
    store = _open(request, key)
    try:
        metadata = store.metadata()
        body, mime, extension = render_chat_export(fmt, store.chat_history(), metadata)
    except ValueError as exc:
        raise HTTPException(
            status_code=422, detail={"error": {"code": "unknown-format", "message": str(exc)}}
        ) from exc
    finally:
        store.close()

    stem = metadata.started_at.strftime("%Y%m%d-%H%M") if metadata else key
    return Response(
        content=body,
        media_type=mime,
        headers={"Content-Disposition": f'attachment; filename="{stem}-chat.{extension}"'},
    )


def _recording(request: Request, key: str):  # noqa: ANN202 - returns RecordingLayout
    """The recording folder for a session, or 404 saying there is not one."""
    layout = resolve_recording(archive.recording_dir(_config(request)), key)
    if layout is None:
        raise HTTPException(
            status_code=404,
            detail={
                "error": {
                    "code": "no-recording",
                    "message": "There is no recording folder for that session.",
                    "severity": "warning",
                }
            },
        )
    return layout


@router.get("/{key}/export/options")
async def export_options(request: Request, key: str) -> dict[str, Any]:
    """What this recording is, and what each preset would turn it into (D-037).

    **The measurement is the point.** A preset list on its own says "720p"; this says what 720p
    means for *this* file — the reported seminar asked for a 720p ceiling at capture and recorded
    at 2560 x 1532, so a window offering choices against the configured number would be describing
    a recording that does not exist.

    Every size is a range and says so. A constant-quality encode depends on content the estimator
    has not watched, and a figure presented as exact would be wrong in the way that matters, since
    the whole question is whether a file is small enough to send.
    """
    layout = _recording(request, key)
    video = layout.existing_video()
    source = probe_media(video) if video is not None else SourceProfile()

    return {
        "key": key,
        "source": source.as_dict(),
        "presets": [
            {**plan.as_dict(source), "estimate": estimate_export(plan, source).as_dict()}
            for plan in PRESETS
        ],
        "default": DEFAULT_PRESET.id,
        "chat_messages": len(_chat_count(request, key)),
        "job": _running_job(request, key),
    }


@router.post("/{key}/export/start")
async def start_export(
    request: Request,
    key: str,
    preset: str = Query(DEFAULT_PRESET.id),
    include_chat: bool = Query(False),
) -> dict[str, Any]:
    """Begin an export, and return the job to watch it by.

    A request, not a download. An encode is minutes for an hour of talk, and the old synchronous
    route held a browser connection open for the whole of it — tolerable for a stream copy and not
    for this. Progress arrives over the WebSocket; `export/result` collects the file.
    """
    if by_preset_id(preset) is None:
        raise HTTPException(
            status_code=422,
            detail={
                "error": {
                    "code": "unknown-preset",
                    "message": (
                        f"There is no export preset called {preset!r}. "
                        f"Available: {', '.join(plan.id for plan in PRESETS)}."
                    ),
                }
            },
        )

    layout = _recording(request, key)
    video = layout.existing_video()
    if video is None:
        raise HTTPException(
            status_code=409,
            detail={
                "error": {
                    "code": "not-exportable",
                    "message": (
                        "This recording has no video, so there is nothing for the player to show. "
                        "The transcript can still be exported as Markdown, SRT, or JSON."
                    ),
                    "severity": "warning",
                }
            },
        )

    runner = request.app.state.export_runner
    source = probe_media(video)
    job, plan = runner.plan_job(key=key, preset_id=preset, source=source)

    store = _open(request, key)
    metadata = store.metadata()
    path = store.path
    store.close()

    started = runner.start(
        job=job,
        plan=plan,
        source=source,
        session_path=path,
        metadata=metadata,
        layout=layout,
        config=_config(request),
        include_chat=include_chat,
    )
    if not started:
        raise HTTPException(
            status_code=409,
            detail={
                "error": {
                    "code": "export-busy",
                    "message": (
                        "An export is already running. Exports use every core this machine has, "
                        "so they run one at a time."
                    ),
                    "severity": "warning",
                }
            },
        )
    return {"started": True, "job": job.as_event()}


@router.get("/{key}/export/status")
async def export_status(request: Request, key: str) -> dict[str, Any]:
    """The current or most recent export for this session.

    The reconciliation path, exactly as `GET /api/session` is for the live pipeline: the WebSocket
    is how progress arrives and this is how a page that has just loaded finds out where things got
    to without waiting for the next frame.
    """
    return {"key": key, "job": _running_job(request, key)}


@router.post("/{key}/export/cancel")
async def cancel_export(request: Request, key: str) -> dict[str, Any]:
    """Stop an export that is running. The partial output is removed rather than left playable."""
    job = request.app.state.export_jobs.current()
    if job is None or job.key != key:
        return {"cancelled": False}
    job.cancel()
    return {"cancelled": True, "job": job.as_event()}


@router.get("/{key}/export/result")
async def export_result(request: Request, key: str) -> Response:
    """Download what the export produced.

    Separate from the request that started it, so a finished export survives a reload — the archive
    is on disk in the recording's own folder, and this is the route that hands it over.
    """
    job = request.app.state.export_jobs.current()
    if job is None or job.key != key or not job.output_path:
        raise HTTPException(
            status_code=404,
            detail={
                "error": {
                    "code": "no-export",
                    "message": "There is no finished export for that session to download.",
                    "severity": "warning",
                }
            },
        )
    path = Path(job.output_path)
    if not path.is_file():
        raise HTTPException(
            status_code=404,
            detail={
                "error": {
                    "code": "export-gone",
                    "message": "The exported file is no longer on disk. Export it again.",
                    "severity": "warning",
                }
            },
        )
    return FileResponse(
        path,
        media_type="application/zip",
        filename=f"{key}-webapp.zip",
    )


def _running_job(request: Request, key: str) -> dict[str, Any] | None:
    job = request.app.state.export_jobs.current()
    return job.as_event() if job is not None and job.key == key else None


def _chat_count(request: Request, key: str) -> list[Any]:
    store = _open(request, key)
    try:
        return list(store.chat_history())
    finally:
        store.close()


@router.get("/{key}/webapp")
async def export_webapp(request: Request, key: str, include_chat: bool = Query(False)) -> Response:
    """Download this recording as a self-contained HTML web application, in a ZIP.

    Refused rather than degraded when the recording has no video or no transcript. A "web
    application" with an empty player and nothing to read is not the thing that was asked for, and
    shipping one under the same name disappoints quietly instead of explaining — so the refusal
    names which piece is missing and what to do about it.

    Built on a worker thread. A talk is hundreds of megabytes, and copying that much into an archive
    on the event loop stalls every other request including the transcript the user is watching.

    **The archive carries the talk and not the conversation about it**, unless `include_chat` says
    otherwise. The exported page's whole purpose is that whoever receives it connects their own
    model and asks their own questions; opening it to find the panel already half-full of someone
    else's questions, about a talk they have not watched, is the opposite of that.
    """
    config = _config(request)
    layout = resolve_recording(archive.recording_dir(config), key)
    if layout is None:
        raise HTTPException(
            status_code=404,
            detail={
                "error": {
                    "code": "no-recording",
                    "message": "There is no recording folder for that session.",
                    "severity": "warning",
                }
            },
        )

    store = _open(request, key)
    try:
        body = await run_in_threadpool(
            build_webapp,
            key=key,
            store=store,
            metadata=store.metadata(),
            layout=layout,
            config=config,
            include_chat=include_chat,
        )
    except ExportError as exc:
        raise HTTPException(
            status_code=409,
            detail={
                "error": {
                    "code": "not-exportable",
                    "message": str(exc),
                    "severity": "warning",
                }
            },
        ) from exc
    finally:
        store.close()

    return Response(
        content=body,
        media_type="application/zip",
        headers={"Content-Disposition": f'attachment; filename="{key}-webapp.zip"'},
    )


@router.delete("/{key}")
async def delete_session(request: Request, key: str) -> dict[str, Any]:
    """Delete a past session file.

    Refuses the session that is currently recording: deleting the file out from under an open
    connection loses the talk in progress, which is the one thing this application must not do.
    """
    manager = getattr(request.app.state, "session_manager", None)
    if key and key == _running_key(manager):
        raise HTTPException(
            status_code=409,
            detail={
                "error": {
                    "code": "session-running",
                    "message": "That session is still recording. Stop it before deleting it.",
                }
            },
        )

    path = archive.find(_config(request), key)
    if path is None:
        return {"deleted": False}
    path.unlink()
    return {"deleted": True}
